control/pure_pursuit/src/pure_pursuit/pure_pursuit_lateral_controller.cpp

```diff
@@ -157,6 +203,83 @@ void PurePursuitLateralController::setResampledTrajectory()
     motion_utils::convertToTrajectoryPointArray(*trajectory_resampled_));
 }
 
+void PurePursuitLateralController::averageFilterTrajectory(
+  autoware_auto_planning_msgs::msg::Trajectory & u)
+{
+  if (static_cast<int>(u.points.size()) <= 2 * param_.path_filter_moving_ave_num) {
+    RCLCPP_ERROR(node_->get_logger(), "Cannot smooth path! Trajectory size is too low!");
+    return;
+  }
+
+  autoware_auto_planning_msgs::msg::Trajectory filtered_trajectory(u);
+
+  for (int64_t i = 0; i < static_cast<int64_t>(u.points.size()); ++i) {
+    TrajectoryPoint tmp{};
+    int64_t num_tmp = param_.path_filter_moving_ave_num;
+    int64_t count = 0;
+    double yaw = 0.0;
+    if (i - num_tmp < 0) {
+      num_tmp = i;
+    }
+    if (i + num_tmp > static_cast<int64_t>(u.points.size()) - 1) {
+      num_tmp = static_cast<int64_t>(u.points.size()) - i - 1;
+    }
+
+    for (int64_t j = -num_tmp; j <= num_tmp; ++j) {
+      tmp.pose.position.x += u.points.at(static_cast<size_t>(i + j)).pose.position.x;
+      tmp.pose.position.y += u.points.at(static_cast<size_t>(i + j)).pose.position.y;
+      tmp.pose.position.z += u.points.at(static_cast<size_t>(i + j)).pose.position.z;
+      tmp.longitudinal_velocity_mps +=
+        u.points.at(static_cast<size_t>(i + j)).longitudinal_velocity_mps;
+      tmp.acceleration_mps2 += u.points.at(static_cast<size_t>(i + j)).acceleration_mps2;
+      tmp.front_wheel_angle_rad += u.points.at(static_cast<size_t>(i + j)).front_wheel_angle_rad;
+      tmp.heading_rate_rps += u.points.at(static_cast<size_t>(i + j)).heading_rate_rps;
+      yaw += tf2::getYaw(u.points.at(static_cast<size_t>(i + j)).pose.orientation);
+      tmp.lateral_velocity_mps += u.points.at(static_cast<size_t>(i + j)).lateral_velocity_mps;
+      tmp.rear_wheel_angle_rad += u.points.at(static_cast<size_t>(i + j)).rear_wheel_angle_rad;
+      ++count;
```

Comment:
There are many duplicated operations decreasing readability. One suggestion for simplification is like this.

```
tmp.pose.position.x += u.points.at(static_cast<size_t>(i + j)).pose.position.x;
tmp.pose.position.y += u.points.at(static_cast<size_t>(i + j)).pose.position.y;
tmp.pose.position.z += u.points.at(static_cast<size_t>(i + j)).pose.position.z;
tmp.longitudinal_velocity_mps += u.points.at(static_cast<size_t>(i + j)).longitudinal_velocity_mps;
tmp.acceleration_mps2 += u.points.at(static_cast<size_t>(i + j)).acceleration_mps2;
tmp.front_wheel_angle_rad += u.points.at(static_cast<size_t>(i + j)).front_wheel_angle_rad;
tmp.heading_rate_rps += u.points.at(static_cast<size_t>(i + j)).heading_rate_rps;

yaw += tf2::getYaw(u.points.at(static_cast<size_t>(i + j)).pose.orientation);
tmp.lateral_velocity_mps += u.points.at(static_cast<size_t>(i + j)).lateral_velocity_mps;
tmp.rear_wheel_angle_rad += u.points.at(static_cast<size_t>(i + j)).rear_wheel_angle_rad;
```

->

```
const auto & p = u.points.at(static_cast<size_t>(i + j));

tmp.pose.position.x += p.pose.position.x;
tmp.pose.position.y += p.pose.position.y;
tmp.pose.position.z += p.pose.position.z;
tmp.longitudinal_velocity_mps += p.longitudinal_velocity_mps;
tmp.acceleration_mps2 += p.acceleration_mps2;
tmp.front_wheel_angle_rad += p.front_wheel_angle_rad;
tmp.heading_rate_rps += p.heading_rate_rps;

yaw += tf2::getYaw(p.pose.orientation);
tmp.lateral_velocity_mps += p.lateral_velocity_mps;
tmp.rear_wheel_angle_rad += p.rear_wheel_angle_rad;
```

-